src/annotator/plugin/cross-frame.js

```diff
@@ -0,0 +1,112 @@
+// @ts-expect-error - `Plugin` base class is still written in CoffeeScript
+import Plugin from '../plugin';
+
+import AnnotationSync from '../annotation-sync';
+import Bridge from '../../shared/bridge';
+import Discovery from '../../shared/discovery';
+import * as frameUtil from '../util/frame-util';
+import FrameObserver from '../frame-observer';
+
+/**
+ * @typedef {import('../../types/annotator').AnnotationData} AnnotationData
+ */
+
+/**
+ * `CrossFrame` provides a connection from the annotator to the sidebar.
+ *
+ * It can be used to publish events to and subscribe to events from the sidebar.
+ *
+ * This class also has logic for injecting Hypothesis into iframes that
+ * are added to the page if they have the `enable-annotation` attribute set
+ * and are same-origin with the current document.
+ */
+export default class CrossFrame extends Plugin {
+  constructor(element, options) {
+    super(element, options);
+
+    const { config, server, on, emit } = options;
+    const discovery = new Discovery(window, { server });
+    const bridge = new Bridge();
+    const annotationSync = new AnnotationSync(bridge, { on, emit });
+    const frameObserver = new FrameObserver(element);
+    const frameIdentifiers = new Map();
+
+    /**
+     * Inject Hypothesis into a newly-discovered iframe.
+     */
+    const injectToFrame = frame => {
+      if (!frameUtil.hasHypothesis(frame)) {
+        const { clientUrl } = config;
+
+        frameUtil.isLoaded(frame, () => {
+          const subFrameIdentifier = discovery.generateToken();
+          frameIdentifiers.set(frame, subFrameIdentifier);
+          const injectedConfig = Object.assign({}, config, {
+            subFrameIdentifier,
+          });
+
+          frameUtil.injectHypothesis(frame, clientUrl, injectedConfig);
+        });
+      }
+    };
+
+    const iframeUnloaded = frame => {
+      bridge.call('destroyFrame', frameIdentifiers.get(frame));
+      frameIdentifiers.delete(frame);
+    };
+
+    /**
+     * Initiate the connection to the sidebar.
+     */
+    this.pluginInit = () => {
```

Comment:
Why is this named `pluginInit`?

Reply:
This overrides a method defined in the `Plugin` base class and called from `Guest`. I forget why `pluginInit` exists rather than just doing the work in the constructor - we might be able to simplify this as part of the re-architecture of the annotator code.